packages/server/utils/atlassian/jiraImages.ts

```diff
@@ -0,0 +1,111 @@
+import cheerio from 'cheerio'
+import crypto from 'crypto'
+import fetch from 'node-fetch'
+import Redis from 'ioredis'
+import ms from 'ms'
+import getRedis from '../getRedis'
+import PROD from '../../PROD'
+
+export const NO_IMAGE_BUFFER = Buffer.from('X')
+export const IMAGE_TTL_MS = ms('2h')
+export const JIRA_IMAGES_ENDPOINT = '/jira-attachemenets'
+
+type UpdateJiraImagesResult = {
+  updatedDescription: string
+  imageUrlToHash: Map<string, string>
+}
+
+/**
+ * Parses a JIRA issue description and replaces the image urls with Parabol's image urls
+ * @param {string} descriptionHTML - HTML string of related Jira issue description
+ * @returns {UpdateJiraImagesResult} - Map of orginal image urls to hashed image urls and updated description
+ */
+export const updateJiraImageUrls = (
+  cloudId: string,
+  descriptionHTML: string
+): UpdateJiraImagesResult => {
+  const imageUrlToHash: Map<string, string> = new Map()
+  const projectBaseUrl = `https://api.atlassian.com/ex/jira/${cloudId}`
+
+  const $ = cheerio.load(descriptionHTML)
+  $('body')
+    .find('img')
+    .each((_i, img) => {
+      const imageUrl = $(img).attr('src')
+      if (!imageUrl) return
+
+      const absoluteImageUrl = `${projectBaseUrl}${imageUrl}`
+      const hashedImageUrl = createImageUrlHash(absoluteImageUrl)
+      imageUrlToHash.set(absoluteImageUrl, hashedImageUrl)
+
+      $(img).attr('src', createParabolImageUrl(hashedImageUrl))
+    })
+
+  return {updatedDescription: $.html(), imageUrlToHash}
+}
+
+export const downloadAndCacheImages = async (
+  authToken: string,
+  imageUrlToHash: Map<string, string>
+) => {
+  const redis = getRedis()
+  const promises: Promise<void>[] = []
+  for (const [imageUrl, hashedImageUrl] of imageUrlToHash) {
+    const cachedImage = await redis.get(hashedImageUrl)
+    if (cachedImage) return // skip if already cached
+
+    promises.push(downloadAndCacheImage(redis, authToken, hashedImageUrl, imageUrl))
+  }
+
+  await Promise.all(promises)
+}
+
+const createParabolImageUrl = (hashedImageUrl: string): string => {
+  //locally API lives on different port than the frontend application
+  return `${
```

Comment:
That's awesome. I was looking for how other endpoints are handled in the project but I couldn't find it. Fixed 👍

Reply:
Ooooh. Interesting. After adding this to `dev.client.config`:
```
      '/jira-attachements': {
        target: `http://localhost:${SOCKET_PORT}`
      }
```
Image request couldn't reach the handler and I started digging what could be the issue.
Seems like hash created that way:
```
const createImageUrlHash = (imageUrl: string) => {
  return crypto
    .createHmac('sha256', serverSecret)
    .update(imageUrl)
    .digest('base64')
}
```
Isn't safe to use in URLs 🤯 . It creates from`https://api.atlassian.com/ex/jira/5bf7b2f4-cec7-4f8a-85f3-5643792a9052/secure/attachment/10000/10000_Screenshot+2021-07-19+at+08.25.20.png` ->`/redZ5TBVrI8sW7FgP79FTY+utdAdaAR/o0raZUkfXc=` and because the hash contains `/` at the beginning it has to be URL encoded. 
More reading: https://www.lytzen.name/2019/09/25/can-a-hash-have-a-slash.html and https://en.wikipedia.org/wiki/Base64#The_URL_applications 

I ended up creating a little helper:
```
const toBase64url = (base64: string) => {
  return base64
    .replace(/=/g, '')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
}
```
And I simply fix the hashed value
```
const createImageUrlHash = (imageUrl: string) => {
  return toBase64url(
    crypto
      .createHmac('sha256', serverSecret)
      .update(imageUrl)
      .digest('base64')
  )
}
```
After that, hash looks like this `_redZ5TBVrI8sW7FgP79FTY-utdAdaAR_o0raZUkfXc` instead of `/redZ5TBVrI8sW7FgP79FTY+utdAdaAR/o0raZUkfXc=`

Reply:
ahh good catch!
we have a package for exactly that purpose. you can see it in action here: https://github.com/ParabolInc/parabol/blob/8e68c8c8106a24333522b25d5ad3306c4de34d5c/packages/server/email/createEmailVerification.ts#L21

so we can rewrite this to something like: 
```ts
base64url.encode(createHmac(...).update(imageUrl))
```

Reply:
perfect! fixed 👍